File: skyvern/library/skyvern_browser_page.py
# ruff: noqa: E402
import asyncio
from contextvars import Context
from typing import TYPE_CHECKING, Any

from skyvern.exceptions import require_local_extra_modules

require_local_extra_modules("skyvern.library.skyvern_browser_page")

from playwright.async_api import Frame, Page

from skyvern.core.script_generations.skyvern_page import SkyvernPage
from skyvern.forge.sdk.api.files import local_file_requires_organization
from skyvern.library.skyvern_browser_page_agent import SkyvernBrowserPageAgent
from skyvern.library.skyvern_browser_page_ai import SdkSkyvernPageAi

if TYPE_CHECKING:
    from skyvern.library.skyvern_browser import SkyvernBrowser


class SkyvernBrowserPage(SkyvernPage):
    """A browser page wrapper that combines Playwright's page API with Skyvern's AI capabilities.

    This class provides a unified interface for both traditional browser automation (via Playwright)
    and AI-powered task execution (via Skyvern). It exposes standard page methods like click, fill,
    goto, etc., while also providing access to Skyvern's task and workflow execution through the
    `run` attribute.

    Example:
        ```python
        # Use standard Playwright methods
        await page.goto("https://example.com")
        await page.fill("#username", "user@example.com")
        await page.click("#login-button")

        # Or use Skyvern's AI capabilities
        await page.agent.run_task("Fill out the contact form and submit it")
        ```

    Attributes:
        agent: SkyvernBrowserPageAgent instance for executing AI-powered tasks and workflows.
    """

    def __init__(self, browser: "SkyvernBrowser", page: Page):
        super().__init__(page, SdkSkyvernPageAi(browser, page))
        self._browser = browser
        self._file_organization_id: str | None = None
        self.agent = SkyvernBrowserPageAgent(browser, page)

    async def _get_file_organization_id(self, file_url: str) -> str | None:
        organization_id = await super()._get_file_organization_id(file_url)
        if organization_id:
            return organization_id
        if not local_file_requires_organization(file_url):
            return None
        if self._file_organization_id is not None:
            return self._file_organization_id
        client = self._browser.skyvern
        if not client._api_key and client._embedded_client is None:
            return None
        # Embedded middleware must not inherit or mutate the caller's context object.
        response = await asyncio.create_task(
            client._client_wrapper.httpx_client.request("api/v1/organizations/me", method="GET"),
            context=Context(),
        )
        response.raise_for_status()
        organization_id = response.json().get("organization_id")
        if not isinstance(organization_id, str) or not organization_id:
            raise PermissionError("Authenticated organization response has no organization ID")
        self._file_organization_id = organization_id
        return organization_id

    async def frame_switch(
        self,
        *,
        selector: str | None = None,
        name: str | None = None,
        index: int | None = None,
    ) -> dict[str, Any]:
        """Switch the working context to an iframe. Exactly one parameter required.

        Args:
            selector: CSS selector for the iframe element (uses content_frame()).
            name: Frame name attribute.
            index: Frame index in page.frames list (0 = main frame).

        Returns:
            Dict with frame name, url, and the parameter used to switch.
        """
        params = sum(p is not None for p in (selector, name, index))
        if params != 1:
            raise ValueError("Exactly one of selector, name, or index is required.")

        frame: Frame | None = None

        if selector is not None:
            element = await self.page.query_selector(selector)
            if element is None:
                raise ValueError(f"Selector '{selector}' did not match any element.")
            frame = await element.content_frame()
            if frame is None:
                raise ValueError(f"Selector '{selector}' did not resolve to an iframe.")

        elif name is not None:
            frame = self.page.frame(name=name)
            if frame is None:
                raise ValueError(f"No frame found with name '{name}'.")

        elif index is not None:
            frames = self.page.frames
            if index < 0 or index >= len(frames):
                raise ValueError(f"Frame index {index} out of range (0-{len(frames) - 1}).")
            frame = frames[index]

        self._working_frame = None if frame is self.page.main_frame else frame
        return {
            "name": frame.name if frame else None,
            "url": frame.url if frame else None,
            "selector": selector,
            "frame_name": name,
            "index": index,
        }

    def frame_main(self) -> dict[str, str]:
        """Switch back to the main page frame, clearing the working iframe."""
        self._working_frame = None
        return {"status": "switched_to_main_frame"}

    async def frame_list(self) -> list[dict[str, Any]]:
        """List all frames on the current page.

        Returns:
            List of dicts with name, url, is_main, and index for each frame.
        """
        frames = self.page.frames
        return [
            {
                "index": i,
                "name": f.name,
                "url": f.url,
                "is_main": f == self.page.main_frame,
            }
            for i, f in enumerate(frames)
        ]

    async def act(
        self,
        prompt: str,
        skip_refresh: bool = False,
        use_economy_tree: bool = False,
    ) -> None:
        """Perform an action on the page using AI based on a natural language prompt.

        Args:
            prompt: Natural language description of the action to perform.

        Examples:
            ```python
            # Simple action
            await page.act("Click the login button")
            ```
        """
        return await self._ai.ai_act(prompt, skip_refresh=skip_refresh, use_economy_tree=use_economy_tree)

File: skyvern-ts/client/src/library/SkyvernBrowserPage.ts
import type { Page } from "playwright";
import type { SkyvernBrowser } from "./SkyvernBrowser.js";
import { SkyvernBrowserPageAgent } from "./SkyvernBrowserPageAgent.js";
import { SkyvernBrowserPageAi } from "./SkyvernBrowserPageAi.js";

/**
 * A browser page wrapper that combines Playwright's page API with Skyvern's AI capabilities.
 *
 * This class provides a unified interface for both traditional browser automation (via Playwright)
 * and AI-powered task execution (via Skyvern). It exposes standard page methods like click, fill,
 * goto, etc., while also providing access to Skyvern's task and workflow execution through the
 * `agent` attribute.
 *
 * @example
 * ```typescript
 * // Use standard Playwright methods
 * await page.goto("https://example.com");
 * await page.fill("#username", "user@example.com");
 * await page.click("#login-button");
 *
 * // Or use Skyvern's AI capabilities
 * await page.agent.runTask("Fill out the contact form and submit it");
 * ```
 */
export class SkyvernBrowserPageCore {
    private readonly _browser: SkyvernBrowser;
    private readonly _page: Page;
    private readonly _ai: SkyvernBrowserPageAi;
    public readonly agent: SkyvernBrowserPageAgent;
    private readonly _proxy: SkyvernBrowserPage;

    private constructor(browser: SkyvernBrowser, page: Page) {
        this._browser = browser;
        this._page = page;
        this._ai = new SkyvernBrowserPageAi(browser, page);
        this.agent = new SkyvernBrowserPageAgent(browser, page);

        this._proxy = new Proxy(this, {
            get(target, prop, receiver) {
                if (prop in target) {
                    return Reflect.get(target, prop, receiver);
                }

                const value = Reflect.get(target._page, prop, target._page);
                if (typeof value === "function") {
                    return value.bind(target._page);
                }
                return value;
            },
        }) as unknown as SkyvernBrowserPage;
    }

    static create(browser: SkyvernBrowser, page: Page): SkyvernBrowserPage {
        const instance = new SkyvernBrowserPageCore(browser, page);
        return instance._proxy;
    }

    get page(): Page {
        return this._page;
    }

    get browser(): SkyvernBrowser {
        return this._browser;
    }

    /**
     * Click an element using a CSS selector, AI-powered prompt matching, or both.
     *
     * This method supports three modes:
     * - **Selector-based**: Click the element matching the CSS selector
     * - **AI-powered**: Use natural language to describe which element to click
     * - **Fallback mode**: Try the selector first, fall back to AI if it fails
     *
     * @param selector - CSS selector for the target element.
     * @param options - Click options including prompt.
     * @param options.prompt - Natural language description of which element to click.
     *
     * @example
     * ```typescript
     * // Click using a CSS selector
     * await page.click("#open-invoice-button");
     *
     * // Click using AI with natural language
     * await page.click({ prompt: "Click on the 'Open Invoice' button" });
     *
     * // Try selector first, fall back to AI if selector fails
     * await page.click("#open-invoice-button", { prompt: "Click on the 'Open Invoice' button" });
     * ```
     */
    async click(selector: string, options?: Parameters<Page["click"]>[1]): Promise<void>;
    async click(options: { prompt: string } & Partial<Parameters<Page["click"]>[1]>): Promise<void>;
    async click(
        selector: string,
        options: { prompt: string } & Partial<Parameters<Page["click"]>[1]>,
    ): Promise<void>;
    async click(
        selectorOrOptions?: string | ({ prompt: string } & Partial<Parameters<Page["click"]>[1]>),
        options?: Parameters<Page["click"]>[1] | ({ prompt?: string } & Partial<Parameters<Page["click"]>[1]>),
    ): Promise<void> {
        let selector: string | undefined;
        let prompt: string | undefined;
        let clickOptions: Partial<Parameters<Page["click"]>[1]> = {};
        let timeout: number | undefined;

        // Parse arguments
        if (typeof selectorOrOptions === "string") {
            selector = selectorOrOptions;
            if (options && typeof options === "object") {
                const { prompt: p, timeout: t, ...rest } = options as {
                    prompt?: string;
                    timeout?: number;
                } & Partial<Parameters<Page["click"]>[1]>;
                prompt = p;
                timeout = t;
                clickOptions = rest;
            } else if (options) {
                clickOptions = options;
            }
        } else if (selectorOrOptions && typeof selectorOrOptions === "object") {
            const { prompt: p, timeout: t, ...rest } = selectorOrOptions;
            prompt = p;
            timeout = t;
            clickOptions = rest;
        }

        if (!selector && !prompt) {
            throw new Error("Missing input: pass a selector and/or a prompt.");
        }

        // Try to click the element with the original selector first
        let errorToRaise: Error | undefined;
        if (selector) {
            try {
                await this._page.click(selector, { ...clickOptions, timeout });
                return;
            } catch (error) {
                errorToRaise = error as Error;
                selector = undefined;
            }
        }

        // If the original selector doesn't work, try to click the element with the AI generated selector
        if (prompt) {
            await this._ai.aiClick({
                intention: prompt,
                data: Object.keys(clickOptions).length > 0 ? clickOptions : undefined,
                timeout,
            });
            return;
        }

        if (errorToRaise) {
            throw errorToRaise;
        }
    }

    /**
     * Fill an input field using a CSS selector, AI-powered prompt matching, or both.
     *
     * This method supports three modes:
     * - **Selector-based**: Fill the input field with a value using CSS selector
     * - **AI-powered**: Use natural language prompt (AI extracts value from prompt or uses provided value)
     * - **Fallback mode**: Try the selector first, fall back to AI if it fails
     *
     * @param selector - CSS selector for the target input element.
     * @param value - The text value to input into the field.
     * @param options - Fill options including prompt.
     * @param options.prompt - Natural language description of which field to fill and what value.
     *
     * @example
     * ```typescript
     * // Fill using selector and value
     * await page.fill("#email-input", "user@example.com");
     *
     * // Fill using AI with natural language
     * await page.fill({ prompt: "Fill 'user@example.com' in the email address field" });
     *
     * // Try selector first, fall back to AI if selector fails
     * await page.fill("#email-input", "user@example.com", { prompt: "Fill the email address" });
     * ```
     */
    async fill(selector: string, value: string, options?: Parameters<Page["fill"]>[2]): Promise<void>;
    async fill(options: { prompt: string; value?: string } & Partial<Parameters<Page["fill"]>[2]>): Promise<void>;
    async fill(
        selector: string,
        value: string,
        options: { prompt: string } & Partial<Parameters<Page["fill"]>[2]>,
    ): Promise<void>;
    async fill(
        selectorOrOptions?: string | ({ prompt: string; value?: string } & Partial<Parameters<Page["fill"]>[2]>),
        value?: string,
        options?: Parameters<Page["fill"]>[2] | ({ prompt?: string } & Partial<Parameters<Page["fill"]>[2]>),
    ): Promise<void> {
        let selector: string | undefined;
        let fillValue: string | undefined;
        let prompt: string | undefined;
        let fillOptions: Partial<Parameters<Page["fill"]>[2]> = {};
        let timeout: number | undefined;

        if (typeof selectorOrOptions === "string") {
            selector = selectorOrOptions;
            fillValue = value;
            if (options && typeof options === "object") {
                const { prompt: p, timeout: t, ...rest } = options as {
                    prompt?: string;
                    timeout?: number;
                } & Partial<Parameters<Page["fill"]>[2]>;
                prompt = p;
                timeout = t;
                fillOptions = rest;
            } else if (options) {
                fillOptions = options;
            }
        } else if (selectorOrOptions && typeof selectorOrOptions === "object") {
            const { prompt: p, value: v, timeout: t, ...rest } = selectorOrOptions;
            prompt = p;
            fillValue = v;
            timeout = t;
            fillOptions = rest;
        }

        if (!selector && !prompt) {
            throw new Error("Missing input: pass a selector and/or a prompt.");
        }

        let errorToRaise: Error | undefined;
        if (selector && fillValue !== undefined) {
            try {
                await this._page.fill(selector, fillValue, { ...fillOptions, timeout });
                return;
            } catch (error) {
                errorToRaise = error as Error;
                selector = undefined;
            }
        }

        if (prompt) {
            await this._ai.aiInputText({
                value: fillValue,
                intention: prompt,
                data: Object.keys(fillOptions).length > 0 ? fillOptions : undefined,
                timeout,
            });
            return;
        }

        if (errorToRaise) {
            throw errorToRaise;
        }
    }

    /**
     * Select an option from a dropdown using a CSS selector, AI-powered prompt matching, or both.
     *
     * This method supports three modes:
     * - **Selector-based**: Select the option with a value using CSS selector
     * - **AI-powered**: Use natural language prompt (AI extracts value from prompt or uses provided value)
     * - **Fallback mode**: Try the selector first, fall back to AI if it fails
     *
     * @param selector - CSS selector for the target select/dropdown element.
     * @param value - The option value to select.
     * @param options - Select options including prompt.
     * @param options.prompt - Natural language description of which option to select.
     *
     * @example
     * ```typescript
     * // Select using selector and value
     * await page.selectOption("#country", "us");
     *
     * // Select using AI with natural language
     * await page.selectOption({ prompt: "Select 'United States' from the country dropdown" });
     *
     * // Try selector first, fall back to AI if selector fails
     * await page.selectOption("#country", "us", { prompt: "Select United States from country" });
     * ```
     */
    async selectOption(
        selector: string,
        values: string | string[],
        options?: Parameters<Page["selectOption"]>[2],
    ): Promise<void>;
    async selectOption(
        options: { prompt: string; value?: string } & Partial<Parameters<Page["selectOption"]>[2]>,
    ): Promise<void>;
    async selectOption(
        selector: string,
        values: string | string[],
        options: { prompt: string } & Partial<Parameters<Page["selectOption"]>[2]>,
    ): Promise<void>;
    async selectOption(
        selectorOrOptions?:
            | string
            | ({ prompt: string; value?: string } & Partial<Parameters<Page["selectOption"]>[2]>),
        values?: string | string[],
        options?: Parameters<Page["selectOption"]>[2] | ({ prompt?: string } & Partial<Parameters<Page["selectOption"]>[2]>),
    ): Promise<void> {
        let selector: string | undefined;
        let selectValue: string | string[] | undefined;
        let prompt: string | undefined;
        let selectOptions: Partial<Parameters<Page["selectOption"]>[2]> = {};
        let timeout: number | undefined;

        // Parse arguments
        if (typeof selectorOrOptions === "string") {
            selector = selectorOrOptions;
            selectValue = values;
            if (options && typeof options === "object") {
                const { prompt: p, timeout: t, ...rest } = options as {
                    prompt?: string;
                    timeout?: number;
                } & Partial<Parameters<Page["selectOption"]>[2]>;
                prompt = p;
                timeout = t;
                selectOptions = rest;
            } else if (options) {
                selectOptions = options;
            }
        } else if (selectorOrOptions && typeof selectorOrOptions === "object") {
            const { prompt: p, value: v, timeout: t, ...rest } = selectorOrOptions;
            prompt = p;
            selectValue = v;
            timeout = t;
            selectOptions = rest;
        }

        if (!selector && !prompt) {
            throw new Error("Missing input: pass a selector and/or a prompt.");
        }

        // Try to select the option with the original selector first
        let errorToRaise: Error | undefined;
        if (selector && selectValue !== undefined) {
            try {
                await this._page.selectOption(selector, selectValue, { ...selectOptions, timeout });
                return;
            } catch (error) {
                errorToRaise = error as Error;
                selector = undefined;
            }
        }

        // If the original selector doesn't work, try to select the option with AI
        if (prompt) {
            await this._ai.aiSelectOption({
                value: typeof selectValue === "string" ? selectValue : selectValue?.[0],
                intention: prompt,
                data: Object.keys(selectOptions).length > 0 ? selectOptions : undefined,
                timeout,
            });
            return;
        }

        if (errorToRaise) {
            throw errorToRaise;
        }
    }

    /**
     * Perform an action on the page using AI based on a natural language prompt.
     *
     * @param prompt - Natural language description of the action to perform.
     *
     * @example
     * ```typescript
     * // Simple action
     * await page.act("Click the login button");
     * ```
     */
    async act(prompt: string): Promise<void> {
        return this._ai.aiAct(prompt);
    }

    async extract(options: {
        prompt: string;
        schema?: Record<string, unknown> | unknown[] | string;
        errorCodeMapping?: Record<string, string>;
        intention?: string;
        data?: string | Record<string, unknown>;
    }): Promise<Record<string, unknown> | unknown[] | string | null> {
        return this._ai.aiExtract(options);
    }

    async validate(prompt: string, model?: Record<string, unknown> | string): Promise<boolean> {
        const normalizedModel: Record<string, unknown> | undefined =
            typeof model === "string" ? { modelName: model } : model;
        return this._ai.aiValidate({ prompt, model: normalizedModel });
    }

    async prompt(
        prompt: string,
        schema?: Record<string, unknown>,
        model?: Record<string, unknown> | string,
    ): Promise<Record<string, unknown> | unknown[] | string | null> {
        const normalizedModel: Record<string, unknown> | undefined =
            typeof model === "string" ? { modelName: model } : model;
        return this._ai.aiPrompt({ prompt, schema, model: normalizedModel });
    }
}

export type SkyvernBrowserPage = SkyvernBrowserPageCore & Page;
